src/shared/completion/use-set-form-completion-mutation.test.js

```diff
@@ -0,0 +1,249 @@
+                wrapper: ({ children }) => (
+                    <Wrapper
+                        queryClientOptions={{ queryCache }}
+                        dataForCustomProvider={{
+                            // never resolving
+                            'dataEntry/dataSetCompletion': () =>
+                                new Promise(() => null),
+                        }}
+                    >
+                        {children}
+                    </Wrapper>
+                ),
+            })
+
+        expect(dataValueSet.current.isLoading).toBe(true)
+
+        // Mutate a current data value
+        act(() => setFormCompletion.current.mutate({ completed: true }))
+
+        await waitForSetFormCompletion(() => {
+            expect(setFormCompletion.current.isLoading).toBe(true)
+        })
+
+        // Ensure that the on-going data value set request is cancelled
+        await waitForDataValueSet(() => {
+            const queryKey = ['dataValues']
+            const cachedDataValuesQuery = queryCache.find({ queryKey })
+            const cachedDataValues = cachedDataValuesQuery.state.data
+
+            expect(cachedDataValues.completeStatus.complete).toBe(true)
+        })
+    })
+
+    it('should revert to the old cache state when an error occurs', async () => {
+        const queryCache = new QueryCache()
+        const onDataValueSuccess = jest.fn()
+        const dataValuesResolver = jest.fn(() => new Promise(() => null))
+
+        // first we need to load the data value set so we have a loading state
+        const { result: dataValueSet, waitFor: waitForDataValueSet } =
+            renderHook(
+                () => useDataValueSet({ onSuccess: onDataValueSuccess }),
+                {
+                    wrapper: ({ children }) => (
+                        <Wrapper
+                            queryClientOptions={{ queryCache }}
+                            dataForCustomProvider={{
+                                dataValues: dataValuesResolver,
+                            }}
+                        >
+                            {children}
+                        </Wrapper>
+                    ),
+                }
+            )
+
+        // Make sure the request is on-going
+        await waitForDataValueSet(() => {
+            expect(dataValuesResolver).toHaveBeenCalledTimes(1)
+            expect(onDataValueSuccess).toHaveBeenCalledTimes(0)
+            return dataValueSet.current.isLoading
+        })
+
+        const { result: setFormCompletion, waitFor: waitForSetFormCompletion } =
+            renderHook(() => useSetFormCompletionMutation(), {
+                wrapper: ({ children }) => (
+                    <Wrapper
+                        queryClientOptions={{ queryCache }}
+                        dataForCustomProvider={{
+                            'dataEntry/dataSetCompletion': () =>
+                                new Promise((_, reject) => {
```

Comment:
could you just return Promise.reject to not have to deal with the fake timers?

Reply:
I'm doing this so I can inspect the query cache before the error occurs.
This way we know that the cache has actually been updated. If that doesn't happen, then `expect(cachedDataValues.completeStatus).toBeUndefined()` could result in a false positive (as the value would be undefined if something else cause the hook to not work for some other, completely unrelated, reason)

Reply:
Got it